### PR TITLE
MSC3083: Add additional tests for joining over federation.

### PR DESCRIPTION
**Pull request by clokep**
This builds on #98 to add tests for matrix-org/synapse#9800.

The tests are pretty much the same, as #98, but the users are no longer on the same server.

### CODE REVIEW

File: tests/msc3083_test.go [hunk 1]
```diff
@@ -127,5 +126,103 @@ func TestRestrictedRoomsLocalJoin(t *testing.T) {
...
+			},
+		},
+	})
+	alice.SendEventSynced(t, space, b.Event{
+		Type:     msc1772SpaceChildEventType,
+		StateKey: &room,
+		Content: map[string]interface{}{
+			"via": []string{"hs1"},
+		},
+	})
```
Comment (richvdh) [inline tests/msc3083_test.go:170]: is it worth factoring this out to a utility function?

File: tests/msc3083_test.go [hunk 2]
```diff
@@ -127,5 +126,103 @@ func TestRestrictedRoomsLocalJoin(t *testing.T) {
...
+
+	// Create a second user and attempt to join the room, it should fail.
+	bob := deployment.Client(t, "hs2", "@bob:hs2")
+	// Note that this is a 400 error because it comes back over federation.
+	FailJoinRoom(bob, t, room, "hs1", 400)
```
Comment (richvdh) [inline tests/msc3083_test.go:175]: can we assert that a useful error message comes back to the client? I'm reluctant to add yet another cause of obscure "no known servers" error messages.

Comment (richvdh) [reply tests/msc3083_test.go:175]: (Indeed, maybe we should spec an error code as part of MSC3083)

File: tests/msc3083_test.go [hunk 3]
```diff
@@ -127,5 +126,103 @@ func TestRestrictedRoomsLocalJoin(t *testing.T) {
...
+	// Join the space, attempt to join the room again, which now should succeed.
+	bob.JoinRoom(t, space, []string{"hs1"})
+	bob.JoinRoom(t, room, []string{"hs1"})
+
+	// Leaving the room works and the user is unable to re-join.
+	bob.LeaveRoom(t, room)
+	bob.LeaveRoom(t, space)
+	FailJoinRoom(bob, t, room, "hs1", 400)
+
+	// Invite the user and joining shuold work.
```
Comment (richvdh) [inline tests/msc3083_test.go:186]:
```suggestion
	// Invite the user and joining should work.
```

File: tests/msc3083_test.go (hunk 1 above)
Comment (clokep) [reply tests/msc3083_test.go:170]: Yeah this part can, the parts with bob cannot since the HS differs between the two tests. :+1:

**Review by Kegsay — APPROVED** [review]
Merge when you're done with this @clokep - LGTM

File: tests/msc3083_test.go (hunk 2 above)
Comment (clokep) [reply tests/msc3083_test.go:153]: I originally had this as a 403 also, but it seems Synapse doesn't properly return federation errors back to clients. I'll look further into it though.

Comment (clokep) [reply tests/msc3083_test.go]: I made this change in Synapse and updated the tests here. This also let us dramatically reduce the copied code between the tests!

I've also updated the spec to mention this error code.

**Review by richvdh — APPROVED** [review]
lgtm!

**Comment by clokep** [comment]
This will also need the changes in matrix-org/synapse#9868 so that the returned error code is correct over federation.

**Comment by clokep** [comment]
This seems to be failing since the latest Docker tag for Synapse doesn't have the required functionality (it will be in Synapse 1.33.0 which should RC this week).

**Comment by clokep** [comment]
There's a new latest image, so hopefully this will pass now. 😄

**Pull request merged**